src/query/api/v1/handler/prometheus/native/read_common.go

```diff
@@ -84,214 +82,63 @@ func read(
 		return emptyResult, err
 	}
 
-	// Block slices are sorted by start time.
-	// TODO: Pooling
-	sortedBlockList := make([]blockWithMeta, 0, initialBlockAlloc)
-	resultChan := result.ResultChan()
-	defer func() {
-		for range resultChan {
-			// NB: drain result channel in case of early termination.
-		}
-	}()
-
-	var (
-		numSteps  int
-		numSeries int
-
-		firstElement bool
-	)
-
-	meta := block.NewResultMetadata()
-	// TODO(nikunj): Stream blocks to client
-	for blkResult := range resultChan {
-		if err := blkResult.Err; err != nil {
-			return emptyResult, err
-		}
-
-		b := blkResult.Block
-		if !firstElement {
-			firstElement = true
-			firstStepIter, err := b.StepIter()
-			if err != nil {
-				return emptyResult, err
-			}
-
-			numSteps = firstStepIter.StepCount()
-			numSeries = len(firstStepIter.SeriesMeta())
-			meta = b.Meta().ResultMetadata
-		}
-
-		// Insert blocks sorted by start time.
-		insertResult, err := insertSortedBlock(b, sortedBlockList,
-			numSteps, numSeries)
-		if err != nil {
-			return emptyResult, err
-		}
-
-		sortedBlockList = insertResult.blocks
-		meta = meta.CombineMetadata(insertResult.meta)
+	res := <-result.ResultChan()
+	if res.Err != nil {
+		return emptyResult, res.Err
```

Comment:
Hm, should we add back the `defer func() { for range resultChan {} }()` before returning here? Otherwise could see it possible to leak goroutines that are blocked tryring to publish to `resultCh`.

Reply:
The channel can only ever have a single value, so that's not actually necessary here